### PR TITLE
Config and DB initialization

### PR DESCRIPTION
**Pull request by viv-defy**

### CODE REVIEW

**Comment by viv-defy** [comment]
### Testing

- Create Config.toml using Config-example.toml
- Add the following code section after line 25, `let last_sync_time = ...` in main.rs
  ```
  println!(
        " {}\n {}\n {}\n {}\n {}\n {}\n {}",
        CONFIG.get_chain(),
        CONFIG.get_native_currency(),
        CONFIG.get_transaction_id_prefix(),
        CONFIG
            .get_last_sync_file_token_transfers()
            .to_string_lossy(),
        CONFIG.get_last_sync_file_transactions().to_string_lossy(),
        CONFIG.get_last_sync_time_token_transfers(),
        CONFIG.get_last_sync_time_transactions()
    );
    ```
- run: `cargo run <table>`
Table can be
- token_transfers
- transactions

File: sync-pipelines-rust/src/settings.rs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+    pub fn get_chain(&self) -> &String {
+        &self.chain
+    }
+
+    pub fn get_native_currency(&self) -> &String {
+        &self.native_currency
+    }
+
+    pub fn get_transaction_id_prefix(&self) -> &String {
```
Comment (ska1296) [inline sync-pipelines-rust/src/settings.rs:66]: Why are you returning a `&String` everywhere? Why not a String or a &str?

File: sync-pipelines-rust/src/settings.rs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    native_currency: String,
+    transaction_id_prefix: String,
+}
+
+const CONFIG_FILE_PATH: &str = "./Config.toml";
+
+impl Settings {
+    pub fn new() -> Result<Self, ConfigError> {
+        let s = Config::builder()
+            .add_source(File::with_name(CONFIG_FILE_PATH).required(false))
```
Comment (ska1296) [inline sync-pipelines-rust/src/settings.rs:36]:
```suggestion
            .add_source(File::with_name(CONFIG_FILE_PATH).required(true))
```
we need a file for config

File: sync-pipelines-rust/src/utils/table.rs [hunk 3]
```diff
@@ -18,7 +20,7 @@ impl Table {
             transfers if transfers == "token_transfers".to_string() => Table::TokenTransfers,
             transactions if transactions == "transactions".to_string() => Table::Transactions,
             _ => {
-                // raise error
+                warn!("Invalid table argument");
```
Comment (ska1296) [inline sync-pipelines-rust/src/utils/table.rs:23]: error, if you are exiting

File: sync-pipelines-rust/src/utils/table.rs [hunk 4]
```diff
@@ -18,7 +20,7 @@ impl Table {
             transfers if transfers == "token_transfers".to_string() => Table::TokenTransfers,
```
Comment (ska1296) [inline sync-pipelines-rust/src/utils/table.rs:20]: why do you need to convert "token_transfers" to_string here?

File: sync-pipelines-rust/src/utils/utils.rs [hunk 5]
```diff
@@ -1,11 +1,26 @@
+use crate::utils::table::Table;
+use crate::CONFIG;
 use chrono::{DateTime, Utc};
 
 pub struct Utils {}
```
Comment (ska1296) [inline sync-pipelines-rust/src/utils/utils.rs:5]: why?

File: sync-pipelines-rust/src/utils/utils.rs [hunk 6]
```diff
@@ -1,11 +1,26 @@
...
 
+const DATETIME_PATTERN: &str = "%Y-%m-%d %H:%M:%S %z";
```
Comment (ska1296) [inline sync-pipelines-rust/src/utils/utils.rs:7]: consider moving it inside the impl block if it is only used here

File: sync-pipelines-rust/src/utils/utils.rs [hunk 7]
```diff
@@ -1,11 +1,26 @@
...
+
 impl Utils {
-    pub fn get_last_synced_time(_table: String) -> DateTime<Utc> {
-        // gets last seen time for the table
-        unimplemented!();
+    pub fn get_last_synced_time(table: Table) -> DateTime<Utc> {
```
Comment (ska1296) [inline sync-pipelines-rust/src/utils/utils.rs:10]: I think it is better to move this to Settings.rs

**Review by ska1296 — CHANGES_REQUESTED** [review]

File: sync-pipelines-rust/config-example.toml [hunk 8]
```diff
@@ -1,12 +1,11 @@
...
-transactions = "~/roverx-rpc/sync-pipelines/transaction_last_sync_blocktimestamp.txt"
-token_transfers = "~/roverx-rpc/sync-pipelines/erc20_last_sync_blocktimestamp.txt"
+token_transfers = "../erc20_last_sync_blocktimestamp.txt"
+transactions = "../transaction_last_sync_blocktimestamp.txt"
 
 [last_sync_time.start_time]
-transactions = 2023-01-01T00:00:00.000Z
-token_transfers = 2023-01-01T00:00:00.000Z
+token_transfers = "2023-01-01 00:00:00 +0000"
+transactions = 0
```
Comment (ska1296) [inline sync-pipelines-rust/config-example.toml:11]: why keep this at 0? It's still an invalid date, right?

**Review by ska1296 — CHANGES_REQUESTED** [review]

File: sync-pipelines-rust/src/utils/utils.rs (hunk 5 above)
Comment (ska1296) [reply sync-pipelines-rust/src/utils/utils.rs:5]: because we are writing impl

File: sync-pipelines-rust/src/utils/utils.rs (hunk 7 above)
Comment (ska1296) [reply sync-pipelines-rust/src/utils/utils.rs:10]: then write would happen in one place and read would happen in another.

File: sync-pipelines-rust/config-example.toml (hunk 8 above)
Comment (viv-defy) [reply sync-pipelines-rust/config-example.toml:11]: Changed transactions' last_sync_time and token_transfers' last_sync_block_number to int type as per the schema: https://github.com/blockchain-etl/ethereum-etl/blob/develop/docs/schema.md#transactionscsv

**Comment by ska1296** [comment]
There are major changes in #391. This PR can be closed and ignored.

**Pull request closed without merging**